packages/matter.js/src/cluster/server/AttributeServer.ts

```diff
@@ -12,22 +12,79 @@ import { Endpoint } from "../../device/Endpoint.js";
 import { Cluster } from "../Cluster.js";
 import { Fabric } from "../../fabric/Fabric.js";
 import { isDeepEqual } from "../../util/DeepEqual.js";
+import { StatusResponseError } from "../../protocol/interaction/InteractionMessenger.js";
+import { StatusCode } from "../../protocol/interaction/InteractionProtocol.js";
+import { MatterError } from "../../common/MatterError.js";
+import { Globals } from "../../model/index.js";
 
+/**
+ * Thrown when an operation cannot complete because fabric information is
+ * unavailable.
+ */
+export class FabricScopeError extends MatterError { }
+
+/**
+ * Base class for all attribute servers.
+ */
 export abstract class BaseAttributeServer<T> {
-    protected value: T;
+    /**
+     * The value is undefined when getter/setter are used. But we still handle the version number here.
+     */
+    protected value: T | undefined = undefined;
     protected version = 0;
     protected endpoint?: Endpoint;
 
     constructor(
         readonly id: number,
         readonly name: string,
         readonly schema: TlvSchema<T>,
-        protected readonly validator: (value: T, name: string) => void,
         readonly isWritable: boolean,
+        readonly isSubscribable: boolean,
         readonly defaultValue: T,
+
+        /**
+         * Optional getter function to handle special requirements or the data are stored in different places.
+         * If a getter method is used for a writable attribute, the setter method must be implemented as well.
+         *
+         * @param session the session that is requesting the value (if any).
+         * @param endpoint the endpoint the cluster server of this attribute is assigned to.
+         * @param isFabricFiltered whether the read request is fabric scoped or not
+         */
+        readonly getter?: (session?: Session<MatterDevice>, endpoint?: Endpoint, isFabricFiltered?: boolean) => T,
+
+        /**
+         * Optional setter function to handle special requirements or the data are stored in different places.
+         * If a setter method is used for a writable attribute, the getter method must be implemented as well.
+         * The method needs to return if the stored value has changed or not.
+         *
+         * @param value the value to be set.
+         * @param session the session that is requesting the value (if any).
+         * @param endpoint the endpoint the cluster server of this attribute is assigned to.
+         * @returns true if the value has changed, false otherwise.
+         */
+        readonly setter?: (value: T, session?: Session<MatterDevice>, endpoint?: Endpoint) => boolean,
```

Comment:
So you mean have devs inherit own classes? In fact the getter/setter/validator are wired in as a commandhandler in the end. That's why a real inheritance is difficult to realize here. Or what do you mean?
Before we had own classes "with a getter" but now with having three would mean a lot of permutations that's why undecided to make it that "flat".

Reply:
No I wasn't thinking actual inheritance...  If you have default handlers that read/write `this.value` then replacing them with handlers in the constructor would just disable the default functionality, but you wouldn't need `if` statements to choose which code path to use.

Also, if the command handler passed to the constructor is implemented using `function()` instead of `=>` it could access `this` and even `super`.  You wouldn't be able to use `super` to access the default implementation though unless it's implemented on the parent class.

Reply:
Ok im understand.I do not like to "too depply connected" idea because I see that as a topic of concerns. It would go too far. but let me try around with the first idea about having default behaviour as default implementation.

Reply:
I did adjustments ... let me know what you think. it is still complex to understand" but much less duplicate code now